services/time/src/lib.rs

```diff
@@ -111,68 +113,98 @@ transactions! {
     }
 }
 
-impl Transaction for TxTime {
-    fn verify(&self) -> bool {
-        self.verify_signature(self.pub_key())
-    }
+#[derive(Debug, Fail)]
+#[repr(u8)]
+enum Error {
+    #[fail(display = "Not authored by a validator")]
+    UnknownSender = 0,
 
-    fn execute(&self, view: &mut Fork) -> ExecutionResult {
-        let validator_keys = Schema::new(&view).actual_configuration().validator_keys;
+    #[fail(display = "The validator time is greater than the proposed one")]
+    ValidatorTimeIsGreater = 1,
+}
+
+impl From<Error> for ExecutionError {
+    fn from(value: Error) -> ExecutionError {
+        ExecutionError::new(value as u8)
+    }
+}
 
-        // The transaction must be signed by the validator.
-        let signed = validator_keys.iter().any(|&validator| {
-            validator.service_key == *self.pub_key()
-        });
+impl TxTime {
+    fn check_signed_by_validator(&self, snapshot: &Snapshot) -> ExecutionResult {
+        let keys = Schema::new(&snapshot).actual_configuration().validator_keys;
+        let signed = keys.iter().any(|k| k.service_key == *self.pub_key());
         if !signed {
-            return Ok(());
+            Err(Error::UnknownSender)?
+        } else {
+            Ok(())
         }
+    }
 
-        let mut schema = TimeSchema::new(view);
+    fn update_validator_time(&self, fork: &mut Fork) -> ExecutionResult {
+        let mut schema = TimeSchema::new(fork);
         match schema.validators_times().get(self.pub_key()) {
             // The validator time in the storage should be less than in the transaction.
-            Some(storage_time) if storage_time >= self.time() => {
-                return Ok(());
-            }
+            Some(time) if time >= self.time() => Err(Error::ValidatorTimeIsGreater)?,
             // Write the time for the validator.
             _ => {
                 schema.validators_times_mut().put(
                     self.pub_key(),
                     self.time(),
-                )
+                );
+                Ok(())
             }
         }
+    }
+
+    fn update_consolidated_time(fork: &mut Fork) {
+        let keys = Schema::new(&fork).actual_configuration().validator_keys;
+        let mut schema = TimeSchema::new(fork);
 
         // Find all known times for the validators.
         let mut validator_times: Vec<SystemTime>;
```

Comment:
I suggest using the following code here:

```rust
let validator_times = {
    let idx = schema.validators_times();
    let mut times: Vec<SystemTime> = idx.iter()
        .filter_map(|(public_key, time)| {
            keys.iter()
                .find(|validator| validator.service_key == public_key)
                .map(|_| time)
        })
        .collect();
    times.sort();
    times.reverse();
    times
};
```

(You may use `sort_by` - it's probably faster, although I find it less understandable).